src/core/reactor_backend.cc

```diff
@@ -1486,10 +1492,98 @@ class reactor_backend_uring final : public reactor_backend {
         delete pfd;
     }
     virtual future<std::tuple<pollable_fd, socket_address>> accept(pollable_fd_state& listenfd) override {
-        return _r.do_accept(listenfd);
+        if (is_speculated(listenfd, POLLIN | POLLOUT)) {
+            listenfd.maybe_no_more_recv();
+            socket_address sa;
+            auto maybe_fd = listenfd.fd.try_accept(sa, SOCK_CLOEXEC);
+            if (maybe_fd) {
+                listenfd.speculate_epoll(EPOLLIN);
+                pollable_fd pfd(std::move(*maybe_fd), pollable_fd::speculation(EPOLLOUT));
+                return make_ready_future<std::tuple<pollable_fd, socket_address>>(std::move(pfd), std::move(sa));
+            }
+        }
+        class accept_completion final : public io_completion {
+            pollable_fd_state& _listenfd;
+            socket_address _sa;
+            promise<std::tuple<pollable_fd, socket_address>> _result;
+        public:
+            accept_completion(pollable_fd_state& listenfd)
+                : _listenfd(listenfd) {}
+            void complete(size_t fd) noexcept final {
+                _listenfd.speculate_epoll(EPOLLIN);
+                pollable_fd pfd(file_desc::from_fd(fd), pollable_fd::speculation(EPOLLOUT));
+                _result.set_value(std::move(pfd), std::move(_sa));
+                delete this;
+            }
+            void set_exception(std::exception_ptr eptr) noexcept final {
+                _result.set_exception(eptr);
+                delete this;
+            }
+            future<std::tuple<pollable_fd, socket_address>> get_future() {
+                return _result.get_future();
+            }
+            ::sockaddr* posix_sockaddr() {
+                return &_sa.as_posix_sockaddr();
+            }
+            socklen_t* socklen_ptr() {
+                return &_sa.addr_length;
+            }
+        };
+        auto desc = std::make_unique<accept_completion>(listenfd);
+        return readable_or_writeable(listenfd).then([&listenfd, desc = std::move(desc), this]() mutable {
```

Comment:
Without speculating here, this will be slower, not faster, since speculation will usually be false, we'll go into the kernel to poll, then go to the kernel again for the accept.

I suggest to drop this patch and come back to it later. Even if it gives a huge speedup it will not be noticeable since only a tiny fraction of the cpu is spent on accepts. Let's concentrate on send/recv first.

Reply:
sure. will create another PR after the send/recv one lands.